src/server/gc_worker/applied_lock_collector.rs

```diff
@@ -0,0 +1,754 @@
+                }
+            };
+
+            if lock.ts <= max_ts {
+                let key = Key::from_encoded_slice(put_request.get_key());
+                locks.push((key, lock));
+            }
+        }
+        if !locks.is_empty() {
+            self.send(locks);
+        }
+    }
+}
+
+impl ApplySnapshotObserver for LockObserver {
+    fn pre_apply_plain_kvs(
+        &self,
+        _: &mut ObserverContext<'_>,
+        cf: CfName,
+        kv_pairs: &[(Vec<u8>, Vec<u8>)],
+    ) {
+        if cf != CF_LOCK {
+            return;
+        }
+
+        if !self.is_clean() {
+            return;
+        }
+
+        let max_ts = self.load_max_ts();
+        if max_ts.is_zero() {
+            return;
+        }
+
+        let locks: Result<Vec<_>> = kv_pairs
+            .iter()
+            .map(|(key, value)| {
+                Lock::parse(value)
+                    .map(|lock| (key, lock))
+                    .map_err(|e| ErrorInner::Mvcc(e.into()).into())
+            })
+            .filter(|result| result.is_err() || result.as_ref().unwrap().1.ts <= max_ts)
+            .map(|result| {
+                // `pre_apply_plain_keys` will be invoked with the data_key in RocksDB layer. So we
+                // need to remove the `z` prefix.
+                result.map(|(key, lock)| (Key::from_encoded_slice(origin_key(key)), lock))
+            })
+            .collect();
+
+        match locks {
+            Err(e) => {
+                error!(
+                    "cannot parse lock";
+                    "err" => ?e
+                );
+                self.mark_dirty()
+            }
+            Ok(l) => self.send(l),
+        }
+    }
+
+    fn pre_apply_sst(&self, _: &mut ObserverContext<'_>, cf: CfName, _path: &str) {
+        if cf == CF_LOCK {
+            error!("cannot collect all applied lock: snapshot of lock cf applied from sst file");
+            self.mark_dirty();
+        }
+    }
+}
+
+struct LockCollectorRunner {
+    observer_state: Arc<LockObserverState>,
+
+    collected_locks: Vec<(Key, Lock)>,
+}
+
+impl LockCollectorRunner {
+    pub fn new(observer_state: Arc<LockObserverState>) -> Self {
+        Self {
+            observer_state,
+            collected_locks: vec![],
+        }
+    }
+
+    fn handle_observed_locks(&mut self, mut locks: Vec<(Key, Lock)>) {
+        if self.collected_locks.len() >= MAX_COLLECT_SIZE {
+            return;
+        }
+
+        if locks.len() + self.collected_locks.len() >= MAX_COLLECT_SIZE {
+            self.observer_state.is_clean.store(false, Ordering::Release);
+            info!("lock collector marked dirty because received too many locks");
+            locks.truncate(MAX_COLLECT_SIZE - self.collected_locks.len());
+        }
+        self.collected_locks.extend(locks);
+    }
+
+    fn start_collecting(&mut self, max_ts: TimeStamp) -> Result<()> {
+        if self.observer_state.max_ts.load(Ordering::Acquire) == max_ts.into_inner() {
```

Comment:
I think only when previous max ts is less than request's max ts, it can start a new round.

Reply:
To prevent receiving a stale request.

Reply:
Yes, you are right